### PR TITLE
More CMFD Helper Functions + Bug fix for issue #1053

### PR DESCRIPTION
**Pull request by shikhar413**
This is another PR related to transitioning CMFD code from Fortran to Python through the C API. This first PR is part of what I foresee will be a three-part sequence, and focuses on a few more helper functions and code that exposes Fortran variables that are used as part of the CMFD algorithm through the C API. PR2 will push the CMFD code written in Python alongside existing CMFD Fortran code, in addition to a bunch of test cases. The goal of PR2 is not to merge it into the develop branch, but instead to look over the code and verify that the results are consistent between Python and Fortran for a variety of test cases. This will also be where performance results between Fortran and Python will be shown. Once the final comments on the Python CMFD code are addressed, a separate PR3 will take the final CMFD python code and replace existing Fortran code so that any CMFD calls will be now be called solely through the C API. In addition, this PR includes a bug fix for issue #1053.

### CMFD Helper Functions

- Pretty straightforward, but exposes openmc_current_batch, openmc_master, entropy_on, and run_CE variables through the C API, since these variables will be used in the Python implementation of CMFD.
- Additionally, I implement `vector_write` in `src/vector_header.F90` (previously marked as `TODO`) so that both the `phi_n` vector and CMFD matrices are outputted to file. This will be useful in PR2 when verifying consistent results, since not only should simulation output results be the same, but the `write_matrices` tag in `cmfd.xml` can also be used to show that output matrix and vector data are also equivalent between Fortran and Python.
-For multigroup CMFD, there is a bug when extracting energy_in data from tally 3 (surface current tally). At each iteration, 12 tallies are extracted at a time, so the bins must be incremented by this amount as well

### CMFD Algorithm Structure

- Initially, when implementing CMFD through the C API, I had split `openmc_next_batch` into `openmc_init_batch`, `openmc_execute_batch`, and `openmc_finalize_batch`, which was called through the C API, with CMFD code injected in between these three calls to retain the order of how CMFD was called through Fortran in the following manner:

```python
while(next_batch_exists):
  # Execute everything in openmc_next_batch before cmfd_init_batch
  openmc_init_batch()
  # Initialize CFMD parameters
  cmfd_init_batch()
  # Execute everything in openmc_next_batch in between cmfd_init_batch and execute_cmfd
  openmc_execute_batch()
  # Execute CMFD for current batch
  execute_cmfd()               
  # Execute everything in openmc_next_batch after execute_cmfd
  openmc_finalize_batch()
```

 However, another option would be to keep `openmc_next_batch` as is, and call `cmfd_init_batch` and `execute_cmfd` before and after this call respectively. This option does technically work, but does not replicate the same order of calls as the Fortran implementation of CMFD. The pseudocode would look as follows:

```python
while(next_batch_exists):
  # Initialize CFMD parameters
  cmfd_init_batch()
  # Run calculations for next batch
  openmc_next_batch()
  # Execute CMFD for batch
  execute_cmfd()               
```

While the second option is appealing in that it does not modify the existing structure of the code, it does not guarantee that the correct values are outputted to statepoint, since statepoint variables are called in `openmc_next_batch` after `execute_cmfd` is called. Since I won't be at the point of handling CMFD runs from restart files by PR3, I can include work on correctly implementing this functionality as part of a future PR. The reason I am including this description here is if there is discussion to go down the route of implementing CMFD through the first option, I can include the additional functions, `openmc_init_batch`, `openmc_execute_batch`, and `openmc_finalize_batch` as part of this PR

### Bug fix for issue #1053 

I am also including a bug fix for issue #1053. This allows simulations to be restarted from statepoint files that were generated in the inactive cycles. The issue is that if a restart run is called from a statepoint that is in the inactive cycles, the subroutine `calculate_average_keff` is not called because the condition `current_batch <= restart_batch` is not met in `openmc_next_batch` (current_batch is equal to restart_batch when loading statepoint file but incremented by one in init_batch right before this conditional). To get around this, I decrement `current_batch` by one in simulation_init so that it starts at the correct location when reaching the conditional. I don't do this for the active cycles, as this would require having to modify `n_realizations` as well. In addition, I also only increment n_realizations in replay_batch_history by one only if replaying in active cycles, as n_realizations should not be incremented in inactive cycles. Finally, I modify the conditional to turn on the inactive timer, as currently it does not start if restarting simulation from inactives and Calculation Rate (inactive) yields infinity. The logic is kind of convoluted but another possible fix would be to not increment current batch by one in `initialize_batch` if part of a restart run and in the inactives

### CODE REVIEW

File: src/cmfd_data.F90 [hunk 1]
```diff
@@ -254,7 +254,7 @@ subroutine compute_xs()
 
                 ! Set the energy bin if needed
                 if (energy_filters) then
-                  filter_matches(i_filter_ein) % bins % data(1) = ng - h + 1
+                  filter_matches(i_filter_ein) % bins % data(1) = 12*(ng - h) + 1
```
Comment (paulromano) [inline src/cmfd_data.F90]: Looking at how this is used, I don't think this is the correct fix actually. The order of bins in the tally will be something like:
```
mesh bin 1, xmin out, group 1
mesh bin 1, xmin out, group 2
...
mesh bin 1, xmin out, group N
mesh bin 2, xmin in, group 1
mesh bin 2, xmin in, group 2
...
```
So the way we are using `score_index` below is just fundamentally wrong. It is assuming that `mesh bin i, xmin out, group j` and `mesh bin i, xmin in, group j` are consecutive in memory, which in fact that are actually separated by the number of energy groups.

File: src/simulation.F90 [hunk 2]
```diff
@@ -503,6 +506,12 @@ function openmc_simulation_init() result(err) bind(C)
       else
         current_batch = restart_batch - n_realizations*n_procs
       end if
+      ! If simulation restarted from inactive batch, decrement current_batch
+      ! by one to replay an additional batch so that keff is set before resuming
+      ! simulation
+      if (restart_batch <= n_inactive) then
+        current_batch = current_batch - 1
```
Comment (paulromano) [inline src/simulation.F90]: Maybe we should just have this check when a statepoint is loaded and set keff then?

**Comment by smharper** [comment]
We also might want to just get rid of the rewind feature since this logic is starting to look a little convoluted.  I'll try to take a look at the code today or tomorrow.

**Comment by paulromano** [comment]
@smharper Agreed, probably better to just remove it. That doesn't necessarily need to be part of this PR though.

**Comment by shikhar413** [comment]
Got it, I can make the fixes as part of this PR. Would getting rid of the rewind feature simply mean removing `replay_batch_history` altogether and making sure all variables are set correctly from `restart_batch`?

**Comment by smharper** [comment]
@shikhar413, correct.  Are you willing to give that a try?  I think that will make for something cleaner and easier to understand (but I don't really know so be sure to chime in if you see some difficulty there).

**Comment by shikhar413** [comment]
Actually this seemed to work, I now set `k_sum`, `keff`, and `current_batch` in `load_source_point`. I still need to add logic to turn on active timer during a restart run, but let me know what you guys think. I think I addressed the CMFD issue as well

**Comment by shikhar413** [comment]
Updated the test suite so that `openmc.capi.next_batch` is only called 3 times instead of 5 times in `test_capi.py` to reflect the fact that we no longer need to replay the batch history from the first active cycle

File: src/cmfd_data.F90 [hunk 3]
```diff
@@ -254,7 +254,7 @@ subroutine compute_xs()
 
                 ! Set the energy bin if needed
                 if (energy_filters) then
-                  filter_matches(i_filter_ein) % bins % data(1) = ng - h + 1
+                  filter_matches(i_filter_ein) % bins % data(1) = ng - h
```
Comment (paulromano) [inline src/cmfd_data.F90]: Are you sure this is right? This means the values will range from ng-1 to 0.

File: src/cmfd_data.F90 [hunk 4]
```diff
@@ -265,39 +265,39 @@ subroutine compute_xs()
 
                 ! Left surface
                 cmfd % current(1,h,i,j,k) = t % results(RESULT_SUM, 1, &
-                     score_index + OUT_LEFT)
+                     score_index + ng*OUT_LEFT)
```
Comment (paulromano) [inline src/cmfd_data.F90]: At this point, it looks to me like `score_index + 1` would correspond to `OUT_LEFT` for the given mesh cell. This means that `score_index + 1 + ng` would correspond to `IN_LEFT`, `score_index + 1 + 2*ng` would correspond to `OUT_RIGHT`, etc. Can you check my logic? If it's right, that means each expression should look something like `score_index + 1 + ng*(z - 1)` where `z` is `OUT_LEFT`, `IN_LEFT`, `OUT_RIGHT`, etc.

Comment (shikhar413) [reply src/cmfd_data.F90]: You're right, my first implementation was specific to the 2 group case but didn't work for anything beyond 2 groups.

**Comment by shikhar413** [comment]
Hmm I just found a new bug with HDF5 while running CMFD on the 2D BEAVRS case. @paulromano I'm going to try to add a fix for that in this PR, if you could hold off on merging the PR for a bit.

**Comment by shikhar413** [comment]
So the HDF5 error was thrown because currently there is no check for whether the CMFD mesh has the same id as another mesh, so I was running a case where the cmfd mesh and entropy mesh had the same id. A fix for this doesn't seem pressing since changing the entropy mesh id avoided the problem, but I can create a mesh_dict to check for duplicate id's in a future PR.

**Comment by paulromano** [comment]
@shikhar413 That's fine. So you're not planning on making any further changes on this PR, correct?

**Comment by shikhar413** [comment]
Thanks @paulromano

**Pull request merged**